File: code/swift/exercises.swift
import Foundation
//1
// Returns the lowercased version of the first element in the array that satisfies the predicate
func firstThenLowerCase(of `for`: [String], satisfying predicate: (String) -> Bool) -> String? {
    `for`.first(where: predicate)?.lowercased()
}

//2
struct Phrase {
    let words: [String]
    
    var phrase: String {
        words.joined(separator: " ")
    }

    func and(_ word: String) -> Phrase {
        // Return a new Phrase thats immutable
        // Append new word
        Phrase(words: words + [word])
    }
}

func say(_ word: String = "") -> Phrase {
    Phrase(words: [word])
}

//3

// Asynchronous line counting function
func meaningfulLineCount(_ filename: String) async -> Result<Int, Error> {
    do {
        let fileURL = URL(fileURLWithPath: filename)
        var count = 0
        for try await line in fileURL.lines {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if !trimmed.isEmpty && !trimmed.starts(with: "#") {
                count += 1
            }
        }
        return .success(count)
    } catch {
        return .failure(error)
    }
}

//4
struct Quaternion: Equatable, CustomStringConvertible {
    let a: Double
    let b: Double
    let c: Double
    let d: Double

    // Default initializer
    init(a: Double = 0, b: Double = 0, c: Double = 0, d: Double = 0) {
        self.a = a
        self.b = b
        self.c = c
        self.d = d
    }

    // Coefficients getter
    var coefficients: [Double] { [a, b, c, d] }

    // Conjugate getter
    var conjugate: Quaternion {
        Quaternion(a: a, b: -b, c: -c, d: -d)
    }

    // Static constants getters
    static let ZERO = Quaternion(a: 0, b: 0, c: 0, d: 0)
    static let I = Quaternion(a: 0, b: 1, c: 0, d: 0)
    static let J = Quaternion(a: 0, b: 0, c: 1, d: 0)
    static let K = Quaternion(a: 0, b: 0, c: 0, d: 1)

    // Addition
    static func + (lhs: Quaternion, rhs: Quaternion) -> Quaternion {
        Quaternion(a: lhs.a + rhs.a,
                   b: lhs.b + rhs.b,
                   c: lhs.c + rhs.c,
                   d: lhs.d + rhs.d)
    }

    // Multiplication Hamilton product
    static func * (lhs: Quaternion, rhs: Quaternion) -> Quaternion {
        Quaternion(
            a: lhs.a * rhs.a - lhs.b * rhs.b - lhs.c * rhs.c - lhs.d * rhs.d,
            b: lhs.a * rhs.b + lhs.b * rhs.a + lhs.c * rhs.d - lhs.d * rhs.c,
            c: lhs.a * rhs.c - lhs.b * rhs.d + lhs.c * rhs.a + lhs.d * rhs.b,
            d: lhs.a * rhs.d + lhs.b * rhs.c - lhs.c * rhs.b + lhs.d * rhs.a
        )
    }

    // String representation (matches test output formatting)
    var description: String {
        // Build terms only for nonzero coefficients
        func term(_ value: Double, _ symbol: String) -> String {
            guard value != 0 else { return "" }
            var result = ""
            if !symbol.isEmpty {
                // include + or - between terms
                if value > 0 { result += "+" }
                // Special case: for ±1, omit the "1"
                if abs(value - 1.0) < 0.0001 {
                    result += symbol
                } else if abs(value + 1.0) < 0.0001 {
                    result += "-\(symbol)"
                } else {
                    result += "\(value)\(symbol)"
                }
            } else {
                // real term (no symbol)
                result += "\(value)"
            }
            return result
        }

        // Construct each component
        let terms = [
            (a, ""),
            (b, "i"),
            (c, "j"),
            (d, "k")
        ]

        // Combine and clean formatting
        var output = ""
        for (_, (value, symbol)) in terms.enumerated() {
            guard value != 0 else { continue }
            let t = term(value, symbol)
            if output.isEmpty {
                // strip leading "+"
                output += t.hasPrefix("+") ? String(t.dropFirst()) : t
            } else {
                output += t
            }
        }

        // Special case for all-zero quaternion
        if output.isEmpty { output = "0" }

        return output
    }
}

//5
indirect enum BinarySearchTree: CustomStringConvertible {
    case empty
    case node(String, BinarySearchTree, BinarySearchTree)

    // Insert (returns a new tree)
    func insert(_ value: String) -> BinarySearchTree {
        switch self {
        case .empty:
            return .node(value, .empty, .empty)
        case let .node(v, left, right):
            if value < v {
                return .node(v, left.insert(value), right)
            } else if value > v {
                return .node(v, left, right.insert(value))
            } else {
                return self // ignore duplicates
            }
        }
    }

    // Contains
    func contains(_ value: String) -> Bool {
        switch self {
        case .empty:
            return false
        case let .node(v, left, right):
            if value == v { return true }
            else if value < v { return left.contains(value) }
            else { return right.contains(value) }
        }
    }

    // Size (computed property)
    var size: Int {
        switch self {
        case .empty:
            return 0
        case let .node(_, left, right):
            return 1 + left.size + right.size
        }
    }

    // String representation
    var description: String {
        switch self {
        case .empty:
            return "()"
        case let .node(v, left, right):
            // Format like "((A)B((C)D))G(H(J)))"
            let l: String
            if case .empty = left { l = "" } else { l = "\(left)" }
            let r: String
            if case .empty = right { r = "" } else { r = "\(right)" }
            return "(\(l)\(v)\(r))"
        }
    }
}

File: code/kotlin/exercises.kt
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException

// 1. firstThenLowerCase - returns the first element satisfying the predicate, lowercased, or null
// Generic function with type constraint: T must be a CharSequence (String, StringBuilder, etc.)
// The 'where' clause restricts the generic type parameter
fun <T> firstThenLowerCase(list: List<T>, predicate: (T) -> Boolean): String? 
    where T : CharSequence {
    // firstOrNull returns null if no match found (safer than first which throws)
    // ?. is the safe call operator - only calls lowercase() if firstOrNull didn't return null
    // This chain returns null if list is empty or no element matches predicate
    return list.firstOrNull(predicate)?.toString()?.lowercase()
}

// 2. say - builder pattern for joining words
// The Say class maintains an immutable list of words and provides a phrase property
// that joins them with single spaces, preserving empty strings for spacing.
class Say(private val words: List<String> = emptyList()) {
    // Custom property getter - computed property that runs this code when accessed
    // 'when' is Kotlin's enhanced switch/if-else expression (returns a value)
    val phrase: String
        get() = when {
            words.isEmpty() -> ""
            words.size == 1 -> words[0]
            else -> {
                // Manually join words with spaces to preserve empty strings correctly
                // Empty strings in the list still contribute to spacing
                val result = StringBuilder()
                // words.indices gives IntRange of valid indices (0..words.size-1)
                for (i in words.indices) {
                    if (i > 0) result.append(" ")
                    result.append(words[i])
                }
                result.toString()
            }
        }
    
    fun and(word: String): Say {
        // List + element creates a new list (immutability)
        // Return a new Say instance with the word appended (immutable pattern)
        return Say(words + word)
    }
}

// Two separate say functions as required (function overloading):
// 1. say() with no parameters - creates an empty Say builder
fun say(): Say {
    return Say()
}

// 2. say(word) with a word parameter - creates a Say builder with the initial word
// listOf() is a standard library function to create immutable lists
fun say(word: String): Say {
    return Say(listOf(word))
}

// 3. meaningfulLineCount - counts non-empty, non-whitespace, non-comment lines
// Returns Long (64-bit integer) as required by the test
fun meaningfulLineCount(filename: String): Long {
    // .use() is Kotlin's try-with-resources equivalent - automatically closes the resource
    // Lambda parameter 'reader' is inferred from context
    BufferedReader(FileReader(filename)).use { reader ->
        // reader.lines() returns a Stream<String> (Java 8+)
        // filter() takes a lambda predicate - only keeps lines that return true
        return reader.lines()
            .filter { line ->
                val trimmed = line.trim()
                // isNotEmpty() is extension function (more readable than length > 0)
                trimmed.isNotEmpty() && !trimmed.startsWith("#")
            }
            .count() // Returns Long
    }
}

// 4. Quaternion - immutable data class for quaternions
// 'data class' automatically generates equals(), hashCode(), toString(), copy(), and component functions
// All properties are 'val' (immutable) - cannot be reassigned after construction
data class Quaternion(
    val a: Double,
    val b: Double,
    val c: Double,
    val d: Double
) {
    // companion object = static members in Java (accessed via Quaternion.ZERO, not instance)
    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }
    
    // Single-expression function syntax - no braces needed when body is a single expression
    fun coefficients(): List<Double> = listOf(a, b, c, d)
    
    fun conjugate(): Quaternion = Quaternion(a, -b, -c, -d)
    
    // operator keyword allows using + and * as operators (q1 + q2 instead of q1.plus(q2))
    // This is operator overloading - enables natural mathematical syntax
    operator fun plus(other: Quaternion): Quaternion {
        return Quaternion(
            a + other.a,
            b + other.b,
            c + other.c,
            d + other.d
        )
    }
    
    operator fun times(other: Quaternion): Quaternion {
        // Quaternion multiplication formula (Hamilton product)
        return Quaternion(
            a * other.a - b * other.b - c * other.c - d * other.d,
            a * other.b + b * other.a + c * other.d - d * other.c,
            a * other.c - b * other.d + c * other.a + d * other.b,
            a * other.d + b * other.c - c * other.b + d * other.a
        )
    }
    
    override fun toString(): String {
        val parts = mutableListOf<String>()
        
        if (a != 0.0) {
            parts.add(formatCoefficient(a))
        }
        
        if (b != 0.0) {
            parts.add(formatImaginary(b, "i"))
        }
        
        if (c != 0.0) {
            parts.add(formatImaginary(c, "j"))
        }
        
        if (d != 0.0) {
            parts.add(formatImaginary(d, "k"))
        }
        
        if (parts.isEmpty()) {
            return "0"
        }
        
        // joinToString with lambda transform - "" means no separator between parts
        // String template: $part inserts the variable value into the string
        return parts.joinToString("") { part ->
            if (part.startsWith("-")) part else if (parts.indexOf(part) == 0) part else "+$part"
        }
    }
    
    private fun formatCoefficient(value: Double): String {
        return if (value == 0.0) {
            "0"
        } else if (value == value.toInt().toDouble()) {
            // For whole numbers, show with .0 (e.g., 1.0 not 1)
            // String template with expression: ${expression} evaluates the expression
            "${value.toInt()}.0"
        } else {
            // String template: $value automatically calls toString()
            "$value"
        }
    }
    
    private fun formatImaginary(value: Double, unit: String): String {
        // 'when' as expression - each branch returns a value
        // No argument to 'when' means it's like a series of if-else conditions
        return when {
            value == 1.0 -> unit
            value == -1.0 -> "-$unit"
            value == value.toInt().toDouble() -> "${value.toInt()}.0$unit"
            else -> "$value$unit"
        }
    }
}

// 5. BinarySearchTree - sealed interface with nested implementations
// 'sealed' restricts inheritance - only classes in this file can implement it
// This enables exhaustive 'when' expressions (compiler knows all possible types)
sealed interface BinarySearchTree {
    fun size(): Int
    fun contains(value: String): Boolean
    fun insert(value: String): BinarySearchTree
    
    // 'object' = singleton instance (like static final in Java)
    // Only one Empty instance exists, accessed as BinarySearchTree.Empty
    object Empty : BinarySearchTree {
        override fun size(): Int = 0
        override fun contains(value: String): Boolean = false
        override fun insert(value: String): BinarySearchTree = Node(value, Empty, Empty)
        override fun toString(): String = "()"
    }
    
    // 'data class' nested inside sealed interface - gets all data class benefits
    // Immutable by default (all properties are 'val')
    data class Node(
        val value: String,
        val left: BinarySearchTree,
        val right: BinarySearchTree
    ) : BinarySearchTree {
        override fun size(): Int = 1 + left.size() + right.size()
        
        override fun contains(value: String): Boolean {
            // 'when' expression - each branch returns a Boolean
            // String comparison uses compareTo() under the hood (< and > operators)
            return when {
                value == this.value -> true
                value < this.value -> left.contains(value)
                else -> right.contains(value)
            }
        }
        
        override fun insert(value: String): BinarySearchTree {
            // Persistent data structure - returns new tree instead of modifying existing
            // Immutability: original tree unchanged, new tree shares unchanged subtrees
            return when {
                value < this.value -> Node(this.value, left.insert(value), right)
                value > this.value -> Node(this.value, left, right.insert(value))
                else -> this // value already exists, return unchanged tree
            }
        }
        
        override fun toString(): String {
            // Smart cast: after 'is Empty' check, compiler knows 'left' is Empty type
            // Can use Empty-specific members without explicit cast
            val leftStr = if (left is Empty) "" else left.toString()
            val rightStr = if (right is Empty) "" else right.toString()
            // String interpolation: variables inserted directly into string
            return "($leftStr$value$rightStr)"
        }
    }
}

